File: 347-top-k-frequent-elements/347-top-k-frequent-elements.cpp
class Solution {
public:
   static bool sortbysec(const pair<int,int> &a,
              const pair<int,int> &b)
{
    return (a.second > b.second);
}
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int,int> umap;
        int n=nums.size();
        for(int i=0;i<n;i++){
            umap[nums[i]]++;
        }
        vector<pair<int,int>> v;
        for(auto it=umap.begin();it!=umap.end();it++){
            v.push_back(make_pair(it->first,it->second));
        }
        
        sort(v.begin(), v.end(), sortbysec);
        // for(auto &i:v)cout<<i.first<<" "<<i.second<<endl;
         vector<int> ans;
        int i=0;
        while(k-- and i<v.size()){
            ans.push_back(v[i].first);
            i++;
        }
        return ans;
        
    }
};

File: 347-top-k-frequent-elements/347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer,Integer> mp = new HashMap<>();
        List<Integer>[] bucket = new List[nums.length+1];
        for(int i=0;i<nums.length;i++){
            mp.put(nums[i],mp.getOrDefault(nums[i],0)+1);
        }
        
        
        for(int i: mp.keySet()){
            int freq = mp.get(i);
            if(bucket[freq]== null){
                bucket[freq]= new ArrayList<>();
            }
            
            bucket[freq].add(i);
        }
        
        
        int[] ans = new int[k];
        int count =0;
        for(int i= bucket.length-1 ;i>=0;i--){
            if(bucket[i]!=null){
                for(int j=0;j<bucket[i].size();j++){
                    ans[count++] = bucket[i].get(j);
                }
            }
            if(count==k)break;
        }
        return ans;
    }
}